compiler/noirc_evaluator/src/ssa/ir/function_inserter.rs

```diff
@@ -34,10 +35,17 @@ impl<'f> FunctionInserter<'f> {
                 super::value::Value::Array { array, typ } => {
                     let array = array.clone();
                     let typ = typ.clone();
-                    let new_array = array.iter().map(|id| self.resolve(*id)).collect();
-                    let new_id = self.function.dfg.make_array(new_array, typ);
-                    self.values.insert(value, new_id);
-                    new_id
+                    let new_array: im::Vector<ValueId> =
+                        array.iter().map(|id| self.resolve(*id)).collect();
+                    if self.const_arrays.get(&new_array) == Some(&value) {
+                        value
```

Comment:
Shouldn't this be `if let Some(value) = self.const_arrays.get(&new_array) {`?
As it is we're only checking if the value in the map exactly matches the original resolved array value. That'd work for when we have one existing array that resolves multiple times but it wouldn't catch the case when we have two identical arrays with different IDs to begin with.

Reply:
Indeed, and this is intentional. If it is the same original value, then it is fine to use the same array.

Reply:
Wouldn't we rather use the value stored in the map though? To deduplicate more arrays

Reply:
It should be correct since arrays are immutable. Maybe it was failing in brillig where arrays are copy-on-write. I suppose it's not important enough to be worth investigating.